pkg/addons/ensure.go

```diff
@@ -413,3 +306,126 @@ func DeleteAddonByName(s *state.State, addonName string) error {
 
 	return errors.Errorf("addon %q does not exist", addonName)
 }
+
+func ensureCSIAddons(s *state.State, addonsToDeploy []addonAction) []addonAction {
+	k8sVersion := semver.MustParse(s.Cluster.Versions.Kubernetes)
+	gte23 := greaterThan23.Check(k8sVersion)
+
+	// We deploy available CSI drivers un-conditionally for k8s v1.23+
+	//
+	// CSIMigration, if applicable, for the cloud providers is turned on by default and requires installation of CSI drviers even if we
+	// don't use external CCM. Although mount operations would fall-back to in-tree solution if CSI driver is not available. Fallback
+	// for provision operations is NOT supported by in-tree solution.
+
+	switch {
+	// CSI driver is required for k8s v.1.23+
+	case s.Cluster.CloudProvider.AWS != nil && (gte23 || s.Cluster.CloudProvider.External):
+		addonsToDeploy = append(addonsToDeploy,
+			addonAction{
+				name: resources.AddonCSIAwsEBS,
+			},
+		)
+	// CSI driver is required for k8s v.1.23+
+	case s.Cluster.CloudProvider.Azure != nil && (gte23 || s.Cluster.CloudProvider.External):
+		addonsToDeploy = append(addonsToDeploy,
+			addonAction{
+				name: resources.AddonCSIAzureDisk,
+			},
+			addonAction{
+				name: resources.AddonCSIAzureFile,
+			},
+		)
+
+	// Install CSI driver unconditionally
+	case s.Cluster.CloudProvider.DigitalOcean != nil:
+		addonsToDeploy = append(addonsToDeploy,
+			addonAction{
+				name: resources.AddonCSIDigitalOcean,
+			},
+		)
+	// Install CSI driver unconditionally
+	case s.Cluster.CloudProvider.Hetzner != nil:
+		addonsToDeploy = append(addonsToDeploy,
+			addonAction{
+				name: resources.AddonCSIHetzner,
+			},
+		)
```

Comment:
Can we also deploy the Nutanix CSI driver here?

```suggestion
		)
	case s.Cluster.CloudProvider.Nutanix != nil:
		addonsToDeploy = append(addonsToDeploy,
			addonAction{
				name: resources.AddonCSINutanix,
			},
		)
```

You would also have to add it here: https://github.com/kubermatic/kubeone/blob/20e645cebb4680fb3b6c3ef22ba82800b5034b6a/pkg/addons/ensure.go#L42

Reply:
Done